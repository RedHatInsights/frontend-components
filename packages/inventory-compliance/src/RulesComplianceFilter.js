import React from 'react';
import propTypes from 'prop-types';
import { FilterDropdown } from '@redhat-cloud-services/frontend-components';
import { HIGH_SEVERITY, MEDIUM_SEVERITY, LOW_SEVERITY } from './Constants';
import { FilterIcon } from '@patternfly/react-icons';

class RulesComplianceFilter extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            hidePassed: props.hidePassed,
            severity: props.severity,
            policy: props.policy,
            filterCategories: [
                {
                    type: 'radio', title: 'Passed', urlParam: 'hidePassed', values: [
                        { label: 'Show all rules', value: false },
                        { label: 'Hide passed rules', value: true }
                    ]
                },
                {
                    type: 'checkbox', title: 'Severity', urlParam: 'severity', values: [
                        { label: HIGH_SEVERITY, value: 'high' },
                        { label: MEDIUM_SEVERITY, value: 'medium' },
                        { label: LOW_SEVERITY, value: 'low' },
                        { label: 'Unknown', value: 'unknown' }
                    ]
                },
                ...props.availablePolicies.length > 1 ? [
                    {
                        type: 'checkbox', title: 'Policy', urlParam: 'policy',
                        values: props.availablePolicies.map(policy => ({
                            label: policy.name, value: policy.name
                        }))
                    }
                ] : []
            ]
        };
    };

    updateInventory = () => {
        const { updateFilter } = this.props;
        const { hidePassed, severity, policy } = this.state;
        updateFilter(hidePassed, severity, policy);
    }

    addFilter = (filterName, selectedValue) => {
        const { severity, policy } = this.state;

        if (filterName === 'hidePassed') {
            this.setState({ hidePassed: selectedValue }, this.updateInventory);
        } else if (filterName === 'severity') {
            this.setState({ severity: [ ...severity, selectedValue ] }, this.updateInventory);
        } else if (filterName === 'policy') {
            this.setState({ policy: [ ...policy, selectedValue ] }, this.updateInventory);
        }
    }

    removeFilter = (filterName, selectedValue) => {
        const { policy, severity } = this.state;

        if (filterName === 'hidePassed') {
            this.setState({ hidePassed: selectedValue }, this.updateInventory);
        } else if (filterName === 'severity') {
            this.setState({ severity: severity.filter((value) => value !== selectedValue) }, this.updateInventory);
        } else if (filterName === 'policy') {
            this.setState({ policy: policy.filter((value) => value !== selectedValue) }, this.updateInventory);
        }
    }

    render() {
        return (
            <FilterDropdown
                id='rules-compliance-filter'
                label={<FilterIcon/>}
                filters = { this.state }
                addFilter={ this.addFilter }
                removeFilter={ this.removeFilter }
                filterCategories={ this.state.filterCategories }
            />
        );
    }
}

RulesComplianceFilter.propTypes = {
    updateFilter: propTypes.func,
    hidePassed: propTypes.bool,
    severity: propTypes.array,
    availablePolicies: propTypes.array,
    policy: propTypes.array
};

RulesComplianceFilter.defaultProps = {
    hidePassed: false,
    severity: [],
    policy: [],
    updateFilter: () => {}
};

export default RulesComplianceFilter;
