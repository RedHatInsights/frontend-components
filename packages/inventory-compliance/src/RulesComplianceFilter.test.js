import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import RulesComplianceFilter from './RulesComplianceFilter';
import { IntlProvider } from 'react-intl';

describe('RulesComplianceFilter component', () => {
    const policies = [
        {
            refId: 'xccdf_org.ssgproject.content_profile_pci-dss',
            name: 'PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 7'
        },
        {
            refId: 'xccdf_org.ssgproject.content_profile_pci-dss',
            name: 'PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 8'
        }
    ];

    it('should render filter', () => {
        const wrapper = shallow(
            <RulesComplianceFilter availablePolicies={ policies } />
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not allow to filter if there is only one policy', () => {
        const wrapper = shallow(
            <RulesComplianceFilter availablePolicies={ [ policies[0] ] } />
        );
        const instance = wrapper.instance();
        expect(wrapper.state('filterCategories').length).toEqual(2);
    });

    it('should allow to add and remove filters', () => {
        const wrapper = shallow(
            <RulesComplianceFilter availablePolicies={ policies } />
        );
        const instance = wrapper.instance();
        instance.addFilter('hidePassed', true);
        expect(wrapper.state('hidePassed')).toEqual(true);
        instance.addFilter('severity', 'medium');
        instance.addFilter('severity', 'low');
        expect(wrapper.state('hidePassed')).toEqual(true);
        expect(wrapper.state('severity')).toEqual([ 'medium', 'low' ]);
        instance.removeFilter('severity', 'low');
        expect(wrapper.state('severity')).toEqual([ 'medium' ]);
        instance.addFilter('policy', policies[0].name);
        instance.addFilter('policy', policies[1].name);
        expect(wrapper.state('policy')).toEqual([ policies[0].name, policies[1].name ]);
        instance.removeFilter('policy', policies[0].name);
        expect(wrapper.state('policy')).toEqual([ policies[1].name ]);
        instance.removeFilter('hidePassed', false);
        expect(wrapper.state('hidePassed')).toEqual(false);
    });
});
