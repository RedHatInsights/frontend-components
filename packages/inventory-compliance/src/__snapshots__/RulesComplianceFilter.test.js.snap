// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RulesComplianceFilter component should render filter 1`] = `
<FilterDropdown
  addFilter={[Function]}
  filterCategories={
    Array [
      Object {
        "title": "Passed",
        "type": "radio",
        "urlParam": "hidePassed",
        "values": Array [
          Object {
            "label": "Show all rules",
            "value": false,
          },
          Object {
            "label": "Hide passed rules",
            "value": true,
          },
        ],
      },
      Object {
        "title": "Severity",
        "type": "checkbox",
        "urlParam": "severity",
        "values": Array [
          Object {
            "label": <React.Fragment>
              <ExclamationCircleIcon
                className="ins-u-failed"
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
                title={null}
              />
               High
            </React.Fragment>,
            "value": "high",
          },
          Object {
            "label": <React.Fragment>
              <ExclamationTriangleIcon
                className="ins-u-warning"
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
                title={null}
              />
               Medium
            </React.Fragment>,
            "value": "medium",
          },
          Object {
            "label": <React.Fragment>
              <svg
                height="1em"
                role="img"
                style={
                  Object {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 18 18"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 0h14a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2V2a2 2 0 0 1 2-2zm6.67 10.46a1.67 1.67 0 1 0 0 3.338 1.67 1.67 0 0 0 0-3.338zm-1.586-6l.27 4.935a.435.435 0 0 0 .434.411H9.55c.232 0 .422-.18.435-.411l.27-4.936A.435.435 0 0 0 9.818 4h-2.3c-.25 0-.448.21-.435.46z"
                  fill="#3A9CA6"
                  fillRule="evenodd"
                />
              </svg>
               Low
            </React.Fragment>,
            "value": "low",
          },
          Object {
            "label": "Unknown",
            "value": "unknown",
          },
        ],
      },
      Object {
        "title": "Policy",
        "type": "checkbox",
        "urlParam": "policy",
        "values": Array [
          Object {
            "label": "PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 7",
            "value": "PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 7",
          },
          Object {
            "label": "PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 8",
            "value": "PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 8",
          },
        ],
      },
    ]
  }
  filters={
    Object {
      "filterCategories": Array [
        Object {
          "title": "Passed",
          "type": "radio",
          "urlParam": "hidePassed",
          "values": Array [
            Object {
              "label": "Show all rules",
              "value": false,
            },
            Object {
              "label": "Hide passed rules",
              "value": true,
            },
          ],
        },
        Object {
          "title": "Severity",
          "type": "checkbox",
          "urlParam": "severity",
          "values": Array [
            Object {
              "label": <React.Fragment>
                <ExclamationCircleIcon
                  className="ins-u-failed"
                  color="currentColor"
                  noVerticalAlign={false}
                  size="sm"
                  title={null}
                />
                 High
              </React.Fragment>,
              "value": "high",
            },
            Object {
              "label": <React.Fragment>
                <ExclamationTriangleIcon
                  className="ins-u-warning"
                  color="currentColor"
                  noVerticalAlign={false}
                  size="sm"
                  title={null}
                />
                 Medium
              </React.Fragment>,
              "value": "medium",
            },
            Object {
              "label": <React.Fragment>
                <svg
                  height="1em"
                  role="img"
                  style={
                    Object {
                      "verticalAlign": "-0.125em",
                    }
                  }
                  viewBox="0 0 18 18"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M2 0h14a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2V2a2 2 0 0 1 2-2zm6.67 10.46a1.67 1.67 0 1 0 0 3.338 1.67 1.67 0 0 0 0-3.338zm-1.586-6l.27 4.935a.435.435 0 0 0 .434.411H9.55c.232 0 .422-.18.435-.411l.27-4.936A.435.435 0 0 0 9.818 4h-2.3c-.25 0-.448.21-.435.46z"
                    fill="#3A9CA6"
                    fillRule="evenodd"
                  />
                </svg>
                 Low
              </React.Fragment>,
              "value": "low",
            },
            Object {
              "label": "Unknown",
              "value": "unknown",
            },
          ],
        },
        Object {
          "title": "Policy",
          "type": "checkbox",
          "urlParam": "policy",
          "values": Array [
            Object {
              "label": "PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 7",
              "value": "PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 7",
            },
            Object {
              "label": "PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 8",
              "value": "PCI-DSS v3 Control Baseline for Red Hat Enterprise Linux 8",
            },
          ],
        },
      ],
      "hidePassed": false,
      "policy": Array [],
      "severity": Array [],
    }
  }
  hideCategories={Array []}
  id="rules-compliance-filter"
  label={
    <FilterIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
      title={null}
    />
  }
  removeFilter={[Function]}
/>
`;
